source/agora/node/QRCodeInterface.d

```diff
@@ -0,0 +1,278 @@
+            temp_kp.secret.toString(PrintMode.Clear),
+            VoterCard(
+                this.key_pair.address,
+                temp_kp.address,
+                expires_time.toISOExtString()
+            )
+        );
+
+        loginInfo.voter_card.signature =
+            this.key_pair.secret.sign(hashFull(loginInfo)[]);
+
+        res.headers["Content-Type"] = "image/svg+xml";
+        res.headers["Vary"] = "Accept-Encoding";
+        res.writeBody(createQRcode(loginInfo));
+    }
+
+    /***************************************************************************
+
+        Respond with QR code containing EncryptionKey.
+
+        Enter the name of the APP & height where the encryptionKey will be used
+        as the get query.
+
+        The encryption key is hashed using the blake2 cryptographic
+        hash function.
+        The encryption key is derived from the pre-image as follows:
+        Hash(pre_image, "APPNAME");
+
+    ***************************************************************************/
+
+    private void keyQRCreate (scope HTTPServerRequest req,
+        scope HTTPServerResponse res)
+    {
+        enforceHTTP("app" in req.query && "height" in req.query,
+        HTTPStatus.badRequest, "Missing app/height field.");
+
+        auto app = req.query["app"];
+        auto height = Height.fromString(req.query["height"]);
+
+        // TODO: PreImage to be released at that height
+        // Or it can be pre-released, so it can be `height + PreimageRevealPeriod`
+        PreImageInfo pre_image;
+
+        // blake2b hashing
+        const Hash key = hashMulti(pre_image, app);
+
+        EncryptionKey encryptionKey = EncryptionKey(
+            app,
+            height,
+            key
+        );
+
+        res.headers["Content-Type"] = "image/svg+xml";
+        res.headers["Vary"] = "Accept-Encoding";
+        res.writeBody(createQRcode(encryptionKey));
+    }
+
+    /***************************************************************************
+
+        Use barcode library to make QR code of JSON format.
+
+        Convert the data structure to JSON and return the QR code image
+        of the SVG format.
+
+        Params:
+            T = Type of struct to QR code
+
+        Returns:
+            QR code of XML string
+
+    ***************************************************************************/
+
+    private string createQRcode (T)(T t)
+        if (is(T == struct))
+    {
+        auto str = serializeToJsonString(t);
+
+        auto svgDrawer = new BaseBarCodeSvgDrawer;
+        svgDrawer.fixSizeMode = true;
+        svgDrawer.W = 400;
+        svgDrawer.H = 400;
+
+        auto qr = new Qr;
+        auto bc = qr.encode(str);
+        return(svgDrawer.draw(bc));
+    }
+}
+
+unittest
+{
+    import std.functional : toDelegate;
+    import vibe.http.client;
+    import vibe.stream.operations;
+    import vibe.core.core;
+    import core.stdc.time;
+    import core.thread;
+    import std.stdio;
```

Comment:
We should **never** use Vibe.d in unittests

Reply:
Remove the Unit Test .